src/ProgramRepresentation/MethodsSet.cpp

```diff
@@ -9,11 +9,13 @@ MethodsSet::MethodsSet(std::set<std::string> methods) {
 
 void MethodsSet::setMethods(std::set<std::string> methods) {
   this->methods.clear();
-  this->methods = methods;
+  this->methods = std::set<std::string>(methods);
```

Comment:
This copy semantics should be documented, both here and for the getter. It's somewhat unintuitive that `MethodsSet::getMethods` doesn't actually return the backing set, but instead returns a copy.

Reply:
I initially did this to be 100% sure a copy was being made since I initially suspected that I might've been returning a reference and that was causing seg faults with the lub function. But now I realize I don't need them, so they'll be removed. The issue had something to do with `std::set_intersection` and `std::set_union.` When I wrote the code for lub like this: 
```
void CalledMethods::leastUpperBound(MethodsSet &preMethods,
                                    MethodsSet &curMethods,
                                    MethodsSet &result) {
  std::set<std::string> res;

  std::set_intersection(preMethods.getMethods().begin(), preSet.getMethods().end(), curSet.getMethods().begin(),
                        curSet.getMethods().end(), std::inserter(res, res.begin()));
  result.setMethods(res);
}
```
I received a seg fault, but changing it to: 
```
void CalledMethods::leastUpperBound(MethodsSet &preMethods,
                                    MethodsSet &curMethods,
                                    MethodsSet &result) {
  std::set<std::string> res;
  std::set<std::string> preSet = preMethods.getMethods();
  std::set<std::string> curSet = curMethods.getMethods();

  std::set_intersection(preSet.begin(), preSet.end(), curSet.begin(),
                        curSet.end(), std::inserter(res, res.begin()));
  result.setMethods(res);
}
```
did not cause me any seg faults. So as far as I can tell, the issue dealt with `set_intersection`/`set_union`, and I didn't find any reference/memory issues elsewhere. But I'm not sure why one version causes a seg fault and the other doesn't, as they should effectively be doing the same thing.

Reply:
Yes, I agree this is really surprising. Can you add a TODO here with a link to your comment saying that we should investigate this at some point - it makes me uncomfortable, but if it's working now I'm okay with moving on.

Reply:
After some research, I realized why the first one seg faults but the second doesn't. It deals with the `.begin` and `.end` iterators.
When I make a call to `getMethods`, it returns a copy of some `set<string>`. In the first one, I'm making two different copies of the `set<string>` by calling std::set_intersection(preMethods.**getMethods**().begin(), preSet.**getMethods**().end() ..., and the beginning and end iterators for these sets likely point to unrelated spots in memory, so it's likely for the program to try to reach inaccessible memory, causing a seg fault. 
For the second one, I'm only making one copy for `preSet` and one copy for `curSet`, so when I call `.begin` and `.end`, the iterators are correct and there won't be a bad memory access.

Reply:
That is such a classic C++ [footgun](https://news.ycombinator.com/item?id=17393292)! Thanks for figuring it out.